### PR TITLE
Docs: update for UDF docs etc.

### PR DESCRIPTION
**Pull request by sk1p**

### CODE REVIEW

File: docs/source/udf/advanced.rst [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+Motivation
+~~~~~~~~~~
+
+Many operations operations can be optimized by working on stacks of frames. For
+example, let's imagine you want to do an elementwise multiplication, which arises
+when doing gain correction.  An implementation in pseudo-Python could look like this
+(for simplicity, we flatten the frame, weights and result to a 1D array here):
+
+.. code-block:: python
```
[the remark this answers is not in the document]
Comment (sk1p) [reply docs/source/udf/advanced.rst]: :+1: - even though these are only supposed to be pseudo-code with Python-syntax, we could make them "real" Python code.

File: docs/source/udf.rst [hunk 2]
```diff
@@ -1,31 +1,111 @@
...
+
+Getting started
+---------------
+
+The easiest way of running a function over your data is using the 
+:meth:`~libertem.api.Context.map` method of the LiberTEM API.
+
+For example, to calculate the sum over the last signal axis:
+
+.. code-block:: python
```
[the remark this answers is not in the document]
Comment (sk1p) [reply docs/source/udf.rst]: :+1:

File: docs/source/concepts.rst [hunk 3]
```diff
@@ -0,0 +1,20 @@
+Concepts
+========
+
+In 4D STEM, an electron beam is scanned over a sample and for each position,
```
[the remark this answers is not in the document]
Comment (sk1p) [reply docs/source/concepts.rst]: Thanks, will add this! This section is for now only linked from the UDF docs; I haven't decied yet where to include it in the documentation hierarchy... Maybe at top level, just before the UDFs?

[the remark this answers is not in the document]
Comment (uellue) [reply docs/source/concepts.rst]: Yes, it should be top-level and relatively early. Conceptually it is related to the architecture, perhaps? Maybe we can do a reorganization step later to manage the growing list of top-level items.

File: docs/source/udf.rst [hunk 4]
```diff
@@ -70,98 +86,268 @@ In pseudocode, data is processed in the following way:
...
+:code:`cloudpickle`.
+
+Note that buffers are only designed to pass lightweight intermediate results
+and thus, it is important that the size of the buffer remains small. Having too
+large buffers can lead to significant decline in performance.
+
+Implementing the processing function
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Now to the actual "meat" of the processing: implementing
```
Comment (uellue) [inline docs/source/udf.rst]: Core instead of meat?

File: docs/source/udf.rst [hunk 5]
```diff
@@ -1,112 +1,353 @@
...
+:class:`~libertem.udf.UDF`. In the simplest case, you need to implement the
+:meth:`~libertem.udf.UDF.get_result_buffers` method and 
+:meth:`~libertem.udf.UDFFrameMixin.process_frame`.
+
+There are two very common patterns for reductions, either reducing over the navigation axes
+into a common accumulator for all frames, keeping the shape of a single frame,
+or reducing over the signal axes and keeping the navigation axes.
+
+A UDF can implement one of these reductions, or even combinations. To handle indexing for you,
+LiberTEM needs to know about the sturcture of your reduction. You can build this structure in the
```
Comment (uellue) [inline docs/source/udf.rst]: sturcture -> structure

File: docs/source/udf/advanced.rst [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+Tiled processing
+----------------
+
+Motivation
+~~~~~~~~~~
+
+Many operations operations can be significantly optimized by working on stacks of frames.
+You can often perform `loop nest optimization <https://en.wikipedia.org/wiki/Loop_nest_optimization>`_
+to improve the `locality of reference <https://en.wikipedia.org/wiki/Locality_of_reference>`_,
+for example using `numba <https://numba.pydata.org/>`_, or using an optimized numpy function.
```
Comment (uellue) [inline docs/source/udf/advanced.rst]: Perhaps add this sentence to show relevance for our target audience?

Gain correction where a constant dark frame is subtracted from each input frame and the result is multiplied with a constant gain map is a typical application in electron microscopy where such optimizations are very effective.

**Pull request merged**